tests/5.0/allocate/test_allocate.c

```diff
@@ -21,7 +21,6 @@ int test_allocate() {
   int* x;
 
 #pragma omp allocate(x) 
-  x = (int *) omp_alloc(N*sizeof(int), omp_null_allocator);
```

Comment:
If you want to access x[0] to x[N], I think you either need `int x[N];` (C array) with `#pragma omp allocate(x)` — or you have to be content that only the pointer gets allocated (`*x`), which does not have room for `N` elements — or you still use `omp_alloc`/`omp_delete` with a pointer.

Reply:
@tob2 I agree too; in my previous comment, the overkill means that the pointer variable is allocated twice: one in "int *x" and the other in "#pragma omp allocate(x)", but no memory allocation occurs for the data to be pointed by the pointer variable. Another way to allocate data to be pointed by the pointer variable is to use an array section as an argument to the allocate directive ("#pragma amp allocate(x[0:N])").

Reply:
@seyonglee :
> Another way to allocate data to be pointed by the pointer variable is to use an array section as an argument to the allocate directive ("#pragma amp allocate(x[0:N])").

I am not quite sure whether that's permitted or not. The current wording is (here for OpenMP 5.2):

"6.5 allocate Directive" - "Arguments" - "list = list of variable list item type"

With "A variable list item is a variable or an array section." – But for something that does has not been allocated yet, "array section" does not make sense (especially not: "The array section must be a subset of the original array." to quote from "3.2.5 Array Sections") .

_I have now filled an OpenMP Spec Issue for this (Issue 3232, not publicly visible)_

Reply:
Bronis pointed out (in that Issue 3232) that "Array Sections" [OpenMP 5.2, p. 66, line 9]) has: “An array section can appear only in clauses for which it is explicitly allowed.”

Thus, "#pragma amp allocate(x[0:N])") is invalid.

Reply:
@tob2 thank you for this clarification. 
I think that I have misunderstood the purpose of the "allocate" directive; so the purpose of the "allocate" directive is to allocate the corresponding "device memory" for the existing host data available on the host memory? Please correct me if there is anything wrong.

Reply:
Thanks @tob2 and @seyonglee for the discussion! This is exactly why we write these tests! :) 
So, in summary, the correct way to go about the allocate directive is:
1) int x[N];
    #pragma omp allocate (x) allocator (any_allocator)
OR for dynamic allocation:
2) int *x;
     #pragma omp allocate (x) allocator (any_allocator)
     x = (int *) omp_alloc(N*sizeof(int), same_any_allocator); //the spec does not mention "same" but that makes sense.
    
I will make sure all these cases are covered in the tests.

Reply:
Thanks @tob2 @spophale; now I have better understanding on the "allocate" directive.